contracts/contracts/strategies/crosschain/AbstractCCTPIntegrator.sol

```diff
@@ -0,0 +1,467 @@
+
+    // Nonce of the last known deposit or withdrawal
+    uint64 public lastTransferNonce;
+
+    mapping(uint64 => bool) private nonceProcessed;
+
+    // For future use
+    uint256[50] private __gap;
+
+    modifier onlyCCTPMessageTransmitter() {
+        require(
+            msg.sender == address(cctpMessageTransmitter),
+            "Caller is not the CCTP message transmitter"
+        );
+        _;
+    }
+
+    constructor(
+        address _cctpTokenMessenger,
+        address _cctpMessageTransmitter,
+        uint32 _destinationDomain,
+        address _destinationStrategy,
+        address _baseToken,
+        address _cctpHookWrapper
+    ) {
+        cctpTokenMessenger = ICCTPTokenMessenger(_cctpTokenMessenger);
+        cctpMessageTransmitter = ICCTPMessageTransmitter(
+            _cctpMessageTransmitter
+        );
+        destinationDomain = _destinationDomain;
+        destinationStrategy = _destinationStrategy;
+        baseToken = _baseToken;
+        cctpHookWrapper = _cctpHookWrapper;
+
+        // Just a sanity check to ensure the base token is USDC
+        uint256 _baseTokenDecimals = Helpers.getDecimals(_baseToken);
+        require(_baseTokenDecimals == 6, "Base token decimals must be 6");
+    }
+
+    function _initialize(uint32 _minFinalityThreshold, uint32 _feePremiumBps)
+        internal
+    {
+        _setMinFinalityThreshold(_minFinalityThreshold);
+        _setFeePremiumBps(_feePremiumBps);
+    }
+
+    function setMinFinalityThreshold(uint32 _minFinalityThreshold)
+        external
+        onlyGovernor
+    {
+        _setMinFinalityThreshold(_minFinalityThreshold);
+    }
+
+    function _setMinFinalityThreshold(uint32 _minFinalityThreshold) internal {
+        // 1000 for fast transfer and 2000 for standard transfer
+        require(
+            _minFinalityThreshold == 1000 || _minFinalityThreshold == 2000,
+            "Invalid threshold"
+        );
+
+        minFinalityThreshold = _minFinalityThreshold;
+        emit CCTPMinFinalityThresholdSet(_minFinalityThreshold);
+    }
+
+    function setFeePremiumBps(uint32 _feePremiumBps) external onlyGovernor {
+        _setFeePremiumBps(_feePremiumBps);
+    }
+
+    function _setFeePremiumBps(uint32 _feePremiumBps) internal {
+        require(_feePremiumBps <= 3000, "Fee premium too high"); // 30%
+
+        feePremiumBps = _feePremiumBps;
+        emit CCTPFeePremiumBpsSet(_feePremiumBps);
+    }
+
+    function handleReceiveFinalizedMessage(
+        uint32 sourceDomain,
+        bytes32 sender,
+        uint32 finalityThresholdExecuted,
+        bytes memory messageBody
+    ) external override onlyCCTPMessageTransmitter returns (bool) {
+        return
+            _handleReceivedMessage(
+                sourceDomain,
+                sender,
+                finalityThresholdExecuted,
+                messageBody
+            );
+    }
+
+    function handleReceiveUnfinalizedMessage(
```

Comment:
I think we might need to implement this, but should we revert when this function gets called. We should only be receiving finalized messages as a result of plain message or standard CCTP token transfer